include/deal.II/matrix_free/dof_info.templates.h

```diff
@@ -1262,39 +1266,102 @@ namespace internal
+                    {
+                      const unsigned int first_i = i;
+                      while (i + 1 < it->second.size() &&
+                             it->second[i + 1] == it->second[i] + 1)
+                        ++i;
+                      range_list.emplace_back(
+                        std::min(it->second[first_i] * chunk_size_zero_vector,
+                                 max_size),
+                        std::min((it->second[i] + 1) * chunk_size_zero_vector,
+                                 max_size));
+                    }
+                  range_list_index[chunk + 1] = range_list.size();
+                }
+              else
+                range_list_index[chunk + 1] = range_list_index[chunk];
+            }
+        };
 
-      vector_zero_range_list_index.resize(
-        1 + task_info
-              .partition_row_index[task_info.partition_row_index.size() - 2],
-        numbers::invalid_unsigned_int);
       std::map<unsigned int, std::vector<unsigned int>> chunk_must_zero_vector;
       for (unsigned int i = 0; i < touched_by.size(); ++i)
         chunk_must_zero_vector[touched_by[i]].push_back(i);
-      vector_zero_range_list.clear();
-      vector_zero_range_list_index[0] = 0;
-      for (unsigned int chunk = 0;
-           chunk < vector_zero_range_list_index.size() - 1;
-           ++chunk)
-        {
-          auto it = chunk_must_zero_vector.find(chunk);
-          if (it != chunk_must_zero_vector.end())
-            {
-              for (const auto i : it->second)
-                vector_zero_range_list.push_back(i);
-              vector_zero_range_list_index[chunk + 1] =
-                vector_zero_range_list.size();
-            }
-          else
-            vector_zero_range_list_index[chunk + 1] =
-              vector_zero_range_list_index[chunk];
-        }
+      const unsigned int n_partitions =
+        task_info.partition_row_index[task_info.partition_row_index.size() - 2];
+      convert_map_to_vector(n_partitions,
+                            chunk_must_zero_vector,
+                            vector_zero_range_list_index,
+                            vector_zero_range_list,
+                            vector_partitioner->local_size());
+
+      // set the import indices in the vector partitioner to one index higher
+      // to indicate that we want to process it first
+      touched_by.resize(
+        (vector_partitioner->local_size() + chunk_size_zero_vector - 1) /
+        chunk_size_zero_vector);
+      for (auto it : vector_partitioner->import_indices())
+        for (unsigned int i = it.first; i < it.second; ++i)
+          touched_by[i / chunk_size_zero_vector] = n_partitions;
+      std::map<unsigned int, std::vector<unsigned int>> chunk_must_do_pre;
+      for (unsigned int i = 0; i < touched_by.size(); ++i)
+        chunk_must_do_pre[touched_by[i]].push_back(i);
+      convert_map_to_vector(n_partitions + 1,
```

Comment:
Why is this `n_partitions +1` while for `chunk_must_zero_vector` the size is `n_partitions`?

Reply:
It is because we need to schedule work on the source *before* the ghost exchange starts, so use another index that is not used within the cell loop. Same for the operation after the loop: We need to indicate that it is done after everything else. (The actual index value does not matter, so I use the same in both cases.)